generators/delimited/main.py

```diff
@@ -0,0 +1,109 @@
+from unicode_csv import UnicodeDictReader
+
+def main():
+    usage = """Origins Datadict Generator.
+
+    Usage:
+        datadict <file_name> [--encoding=<encoding>] [--delimiter=<delimiter>] [--identifier=<indentifier>] [--operation=<assert>|<retract>] [--valid_time=<valid_time>]
+
+    Options:
+        -h --help  Show usage info.
+        --encoding=<encoding>  The text encoding type of the file being read.
+        --delimiter=<delimiter>  The character used to separate items in the csv file.
+        --identifier=<identifier>  A comma-delimited (no spaces) list of fieldnames which make up a unique identifier for an entry in the file. e.g. --identifier=first,last,birthdate
+        --operation=<assert>|<retract>  The origins operation intended for the new facts. (assert|retract)
+        --valid_time=<valid_time>  The valid time for new facts.
+    """ #noqa
+
+    from docopt import docopt
+
+    # Trim 'origins generate' from the line.
+    argv = sys.argv[2:]
+    args = docopt(usage, argv=argv, version='0.1')
+
+   # Defaults
+    encoding = 'utf-8'
+    delimiter = ','
+    operation = 'assert'
+    valid_time = strftime("%Y-%m-%d %H:%M:%S")
+    identifier = None
+
+    # Collect arg values
+    file_name = os.path.join(os.getcwd(), args['<file_name>'])
+    if args['--encoding'] is not None:
+        encoding = args['--encoding']
+    if args['--delimiter'] is not None:
+        delimiter = args['--delimiter']
+    if args['--operation'] is not None:
+        operation = args['--operation']
+    if args['--valid_time'] is not None:
+        valid_time = args['--valid_time']
+    if args['--identifier'] is not None:
+        identifier = args['--identifier'].split(",")
+
+    # No need to validate input because docopt makes sure all arguments are given
+    generate(file_name, encoding, delimiter, identifier, operation, valid_time)
+
+
+# Minimum possible header.
+HEADER = (
+    'operation',
+    'entity',
+    'attribute',
+    'value',
+    'valid_time'
+)
+
+def generate(file_name, encoding, delimiter, identifier, operation, valid_time):
+    # Creating a CSV writer that feeds to stdout.
+    w = csv.writer(sys.stdout)
+
+    # Write the header to start.
+    w.writerow(HEADER)
+
+    # Create each fact
+    with open(file_name, 'rU') as f:
+        sniff = 1024
+        dialect = None
+
+        # Infer various properties about the file
+        # Sample the file to determine the dialect
+        sample = '\n'.join([l for l in f.readlines(sniff)])
+        f.seek(0)
+
+        # Determine dialect
+        sniffer = csv.Sniffer()
+        dialect = sniffer.sniff(sample)
+
+        reader = UnicodeDictReader(f, dialect=dialect, delimiter=delimiter, encoding=encoding)
+
+        for line in reader:
+            # Use the identifier for each entry as the entity for the new fact(s).
+            # Use the remaining fields as attributes for that entity.
+            for key, value in line.items():
+                id = ''
```

Comment:
The logic for building the `id` should be moved to a function and use a list to join the strings:

```python
def build_id(line, fields):
    return '.'.join([line[f] for f in fields])
```

If `identifier` is not defined, default it to `reader.fieldnames[0]` so it does not need to be checked every iteration.